File: retry-reports/python-second/002/20260828-130420/results/storage/data-plane/python/streaming/storage-dp-python-blob-event-notifier/python-azure-skills/azure-skill-mcp/generated-code/event_publisher.py
"""Publish custom downstream events to an Event Grid custom topic."""

from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import Any

from azure.core.exceptions import AzureError
from azure.eventgrid import EventGridEvent

logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class CustomEvent:
    event_type: str
    subject: str
    data: Any
    data_version: str = "1.0"

    def to_event_grid_event(self) -> EventGridEvent:
        if not self.subject.startswith("/"):
            raise ValueError(
                "Custom event subjects must start with '/' to support hierarchy filters"
            )
        return EventGridEvent(
            event_type=self.event_type,
            subject=self.subject,
            data=self.data,
            data_version=self.data_version,
        )


class EventPublisher:
    def __init__(self, publisher_client: Any) -> None:
        self._client = publisher_client

    def publish(self, events: list[CustomEvent]) -> bool:
        if not events:
            return True
        sdk_events = [event.to_event_grid_event() for event in events]
        try:
            self._client.send(sdk_events)
        except AzureError:
            logger.exception("Failed to publish %d downstream event(s)", len(events))
            return False
        logger.info("Published %d downstream event(s)", len(events))
        return True


class AsyncEventPublisher:
    def __init__(self, publisher_client: Any) -> None:
        self._client = publisher_client

    async def publish(self, events: list[CustomEvent]) -> bool:
        if not events:
            return True
        sdk_events = [event.to_event_grid_event() for event in events]
        try:
            await self._client.send(sdk_events)
        except AzureError:
            logger.exception("Failed to publish %d downstream event(s)", len(events))
            return False
        logger.info("Published %d downstream event(s)", len(events))
        return True

File: rerun-reports/lane-a/031/20260828-032156/results/storage/data-plane/java/streaming/storage-dp-java-blob-event-notifier/java-azure-skills/azure-skill-mcp/generated-code/src/main/java/com/example/blobevents/EventPublisher.java
package com.example.blobevents;

import com.azure.core.util.BinaryData;
import com.azure.messaging.eventgrid.EventGridEvent;
import com.azure.messaging.eventgrid.EventGridPublisherClient;
import java.util.List;
import java.util.Objects;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class EventPublisher {
    private static final Logger LOGGER = LoggerFactory.getLogger(EventPublisher.class);

    private final EventGridPublisherClient<EventGridEvent> client;

    public EventPublisher(EventGridPublisherClient<EventGridEvent> client) {
        this.client = Objects.requireNonNull(client, "client");
    }

    public static EventPublisher dryRun() {
        return new EventPublisher();
    }

    private EventPublisher() {
        this.client = null;
    }

    public void publish(List<CustomEvent> events) {
        List<EventGridEvent> sdkEvents = toSdkEvents(events);
        if (client == null) {
            sdkEvents.forEach(event -> LOGGER.info(
                "Dry-run publish: type='{}', subject='{}'",
                event.getEventType(),
                event.getSubject()
            ));
            return;
        }
        client.sendEvents(sdkEvents);
    }

    static List<EventGridEvent> toSdkEvents(List<CustomEvent> events) {
        return events.stream()
            .map(event -> new EventGridEvent(
                event.subject(),
                event.eventType(),
                BinaryData.fromObject(event.data()),
                event.dataVersion()
            ))
            .toList();
    }
}
